### PR TITLE
Update Deseq2 to v. 1.38

### PR DESCRIPTION
**Pull request by tuncK**
FOR CONTRIBUTOR:
* [ ] - I have read the [CONTRIBUTING.md](https://github.com/galaxyproject/tools-iuc/blob/master/CONTRIBUTING.md) document and this tool is appropriate for the tools-iuc repo.
* [ ] - License permits unrestricted use (educational + commercial)
* [ ] - This PR adds a new tool or tool collection
* [x] - This PR updates an existing tool or tool collection
* [ ] - This PR does something else (explain below)

I  wish there would be a way to fix the displayed version numbers to something that follows the version of deseq2 itself. Displaying "2.11.40.8" is misleading.

### CODE REVIEW

**Comment by tuncK** [comment]
Not much change in DeSeq2 itself. I started this after noticing that the DeSeq2 version being used is very hard to figure out for a typical user.

Now addressing: https://github.com/galaxyproject/tools-iuc/issues/3048. I added a boolean switch into the wrapper to turn beta priors on/off.

Looking at https://www.bioconductor.org/packages/release/bioc/manuals/DESeq2/man/DESeq2.pdf, "In
versions >=1.16, the default is set [betaPrior] to FALSE, and shrunken LFCs are obtained
afterwards using lfcShrink". Galaxy wrapper's default is betaPrior = true. We should switch I guess?

**Comment by bernt-matthias** [comment]
> I started this after noticing that the DeSeq2 version being used is very hard to figure out for a typical user.

True .. at least from the tool version. Does the `version_command` is of any help for the user?

> Galaxy wrapper's default is betaPrior = true. We should switch I guess?

IMO yes.

File: tools/deseq2/deseq2.R [hunk 1]
```diff
@@ -37,7 +37,9 @@ options(show.error.messages = FALSE, error = function() {
 })
 
 # we need that to not crash galaxy with an UTF8 error on German LC settings.
-loc <- Sys.setlocale("LC_MESSAGES", "en_US.UTF-8")
+# The line below breaks test #11 and as if not needed anymore?
```
Comment (bgruening) [inline tools/deseq2/deseq2.R:40]: I assume but are not sure, that we added this to fix the `,` vs `.` problem.

Comment (tuncK) [reply tools/deseq2/deseq2.R:40]: Leaving it as is causes "Error in (function (x)  : attempt to apply non-function" in R somehow on my VM.

Comment (bgruening) [reply tools/deseq2/deseq2.R:40]: Interesting, this is unexpected. But yes maybe that is not a problem anymore.

Comment (bgruening) [reply tools/deseq2/deseq2.R:40]: Thanks for including a recent profile!

**Comment by tuncK** [comment]
Addressed https://github.com/galaxyproject/tools-iuc/issues/4425, added an option to provide custom size factors as a tabular from the history.

**Review by bernt-matthias — APPROVED** [review]

File: tools/deseq2/deseq2.xml [hunk 2]
```diff
@@ -105,6 +109,9 @@ Rscript '${__tool_directory__}/deseq2.R'
     #if $advanced_options.auto_mean_filter_off:
         -c
     #end if
+    #if $advanced_options.use_beta_priors:
+        -d
+    #end if
```
Comment (bernt-matthias) [inline tools/deseq2/deseq2.xml:114]: This may be simplified to 

```suggestion
    $advanced_options.use_beta_priors
```

File: tools/deseq2/deseq2.xml [hunk 3]
```diff
@@ -207,28 +224,31 @@ Rscript '${__tool_directory__}/deseq2.R'
...
                 </param>
                 <when value="-P">
-                    <param name="prefilter_value" type="integer" min="0" value="10" label="Pre-filter value" help="Keep only rows that have at least N reads total." />
+                    <param name="prefilter_value" type="integer" min="0" value="10" label="Pre-filter value" help="Keep only rows that have at least N reads total."/>
                 </when>
                 <when value=""/>
             </conditional>
+            <param name="use_beta_priors" type="boolean" truevalue="1" falsevalue="0" checked="false"
+                label="Use beta priors"
+                help="Whether or not to put a zero-mean normal prior on the non-intercept coefficients"/>
```
Comment (bernt-matthias) [inline tools/deseq2/deseq2.xml:233]: If this is changed to 

```
            <param name="use_beta_priors" type="boolean" truevalue="-d" falsevalue="" checked="false"
                label="Use beta priors"
                help="Whether or not to put a zero-mean normal prior on the non-intercept coefficients"/>
```

Or? May be applicable to other booleans as well. 

Implement at will :)

Comment (tuncK) [reply tools/deseq2/deseq2.xml]: Done.

**Comment by tuncK** [comment]
I am stuck with "planemo test" errors and I dont know how to proceed:

1) On my end, 14 out of 14 pass with or without --biocontainers flag.
2) Here on GH, tests #6, 11,12,13 fail with the very same complaint "Error in (function (x)  : attempt to apply non-function"
3) Also on GH, if I let the locale setting to be performed tests #6,7,10,12 fail instead.

Do any of you have any idea what might be off here? That one of the included conda recipes had an unnoticed dependency on something local and we need to rebuild?

**Comment by bernt-matthias** [comment]
Can you call

```
GALAXY_SLOTS=2 planemo test...
```

locally? This should be the only difference...

**Comment by bernt-matthias** [comment]
The container is built  from scratch for each CI run. Maybe you can also try to remove your local image.

Changes in conda are always a possibility.

**Comment by tuncK** [comment]
@bernt-matthias Whatever that irreproducible issue was, is now gone. If you do not have any other suggestions, can you please merge this?

**Comment by tuncK** [comment]
"Incomplete read while reading from https://nodejs.org/download/release/v18.12.1/node-v18.12.1-linux-x64.tar.gz" As if a re-run would fix it?

**Comment by bernt-matthias** [comment]
I restarted it.

**Comment by bgruening** [comment]
restarted and fixed

**Comment by lldelisle** [comment]
The beta parameter change is a big change. We are running a training and using different versions of DESeq2 gives very different results!
How can we warn the users?

**Comment by lldelisle** [comment]
Here is the MA plot with the new default (right) or the beta option to true (like before)
image

**Comment by tuncK** [comment]
Yes, it is supposed to have a significant impact on the results. You can make it behave like in the past by toggling "use beta priors" to true.

**Comment by pavanvidem** [comment]
I did a quick test run with and without the beta priors option on our tutorial data. Yes, the plots look different but the final DE genes set is almost the same. I found 951 out of 954 significant DE genes (`pdj <=0.05`) in common. The main difference is in fold changes which is expected because we do not shrink the fold changes anymore. Previous versions have 113 genes with `abs(FC)>1`. With the current version, we have 204 genes. All 113 genes from the previous version are a subset of the current version. If we increase the FC threshold maybe we get close to the same genes set as before but we need to find a reasonable argumentation for this. Probably `betaPrior=True` makes more sense because, in our Galaxy tool, we do not shrink the fold changes which can be done either with the `lfcShrink` function or `betaPrior=True`. Please refer to this post https://support.bioconductor.org/p/95695/

**Comment by tuncK** [comment]
I have a poor understanding of the theoretical difference between the two sets of p-values. Also here, he says "my current recommendation would be to use the p-values from un-shrunken LFC and then use the shrunken LFC for visualization or ranking of genes. This is the table you get with default DESeq => results => lfcShrink."

My understanding is that one had better introduce the shrinkage function (or even offer multiple options) into deseq2.R, but still report the p-values with beta-prior=False rather than simply recommending the users to set betaPrior=True (which we could also set as the default within the wrapper). If someone could confirm the statistical theory behind it, I would be happy to implement it.

**Comment by pavanvidem** [comment]
Yes, it is even better to call `lfcShrink` after calling `results`. It is also shown in the [vignette](https://bioconductor.org/packages/release/bioc/vignettes/DESeq2/inst/doc/DESeq2.html#differential-expression-analysis) that they used the shrunk FCs for plotting only. I wonder why it was not implemented like this in our wrapper. If we follow the vignette by adding  `lfcShrink` after `results`,  we can remove the `betaPrior` param completely from the Galaxy tool. `DESeq` function generates unshruken FCs by default because `betaPrior` defaults to FALSE. This is the result table we output. Then use the results of `lfcShrink` for plotting.  

In summary, we have to revert back to the version without `betaPrior` option and add `lfcShrink` only before `plotMA` call.

We will still get different results than in the tutorial but the plots will look the same.

**Comment by bernt-matthias** [comment]
> In summary, we have to revert back to the version without betaPrior option and add lfcShrink only before plotMA

I would keep `betaPrior` (it was requested https://github.com/galaxyproject/tools-iuc/issues/3048) and add a  `lfcShrink` call to our R script. 

> We will still get different results than in the tutorial but the plots will look the same.

Thats nice indeed .. but not a strong argument :)

**Comment by tuncK** [comment]
>> In summary, we have to revert back to the version without betaPrior option and add lfcShrink only before plotMA

> I would keep betaPrior (it was requested https://github.com/galaxyproject/tools-iuc/issues/3048) and add a lfcShrink call to our R script.

I think so too. Pavan's suggestion would lead to double-shrink, would not it?

> We will still get different results than in the tutorial but the plots will look the same.

Does anyone here understand the math behind why I am to plot a different p-value rather than the one I am to tabulate? If nobody has a good explanation about it, I think we should contact the authors for clarification.

**Comment by pavanvidem** [comment]
Sure, we can keep it if it was requested. If we set it to FALSE by default, then wouldn't lead to double shrinkage.

The shrinkage applies to MA-plot only which does not plot any p-values but FCs. From the vignette: "It is more useful to visualize the MA-plot for the shrunken log2 fold changes, which remove the noise associated with log2 fold changes from low count genes without requiring arbitrary filtering thresholds.". That makes sense. It affects mainly the genes with a very low mean expression. They usually have very high fold changes which is often just noise. You do not see much difference in FC shrinkage for highly expressed genes.

**Comment by tuncK** [comment]
The FC part is clear to me, p-value is not. Which set of p-values are we going to apply the cutoff's to call DE vs non-DE genes at the end? In the MA-plot above, for instance, you labelled DEs in blue vs gray. But this binary classification is before or after shrinkage and why?

**Comment by pavanvidem** [comment]
I don't think there will be any changes in p-values after shrinkage. The p-values change only when you switch `betaPrior` of `DESeq` function. If you run `DESeq` function followed by  `lfcShrink`, you probably get the same p-values. I haven't tested this though.

**Comment by bernt-matthias** [comment]
Can someone please summarize  this discussion in an issue?

**Comment by pavanvidem** [comment]
Sure, added an issue https://github.com/galaxyproject/tools-iuc/issues/5552
